source/fab/tools/tool_repository.py

```diff
@@ -0,0 +1,142 @@
+        # time the instance is requested (since we overwrite __new__). But
+        # we only want to initialise the instance once, so let the constructor
+        # not do anything if the singleton already exists:
+        if ToolRepository._singleton:
+            return
+
+        self._logger = logging.getLogger(__name__)
+        super().__init__()
+
+        # Create the list that stores all tools for each category:
+        for category in Categories:
+            self[category] = []
+
+        # Add the FAB default tools:
+        # TODO: sort the defaults so that they actually work (since not all
+        # tools FAB knows about are available). For now, disable Fpp:
+        # We get circular dependencies if imported at top of the file:
+        # pylint: disable=import-outside-toplevel
+        from fab.tools import (Ar, Cpp, CppFortran, Gcc, Gfortran,
+                               Icc, Ifort, Psyclone, Rsync)
+
+        for cls in [Gcc, Icc, Gfortran, Ifort, Cpp, CppFortran,
+                    Fcm, Git, Subversion, Ar, Psyclone, Rsync]:
+            self.add_tool(cls)
+
+    def add_tool(self, cls: Type[Any]):
+        '''Creates an instance of the specified class and adds it
+        to the tool repository.
+
+        :param cls: the tool to instantiate.
+        '''
+
+        # Note that we cannot declare `cls` to be `Type[Tool]`, since the
+        # Tool constructor requires arguments, but the classes used here are
+        # derived from Tool which do not require any arguments (e.g. Ifort)
+
+        tool = cls()
+        # We do not test if a tool is actually available. The ToolRepository
+        # contains the tools that FAB knows about. It is the responsibility
+        # of the ToolBox to make sure only available tools are added.
+        self[tool.category].append(tool)
+
+        # If we have a compiler, add the compiler as linker as well
+        if tool.is_compiler:
+            linker = Linker(name=f"linker-{tool.name}", compiler=tool)
+            self[linker.category].append(linker)
+
+    def get_tool(self, category: Categories, name: str) -> Tool:
+        ''':returns: the tool with a given name in the specified category.
+
+        :param category: the name of the category in which to look
+            for the tool.
+        :param name: the name of the tool to find.
+
+        :raises KeyError: if there is no tool in this category.
+        :raises KeyError: if no tool in the given category has the
+            requested name.
+        '''
+
+        if category not in self:
+            raise KeyError(f"Unknown category '{category}' "
+                           f"in ToolRepository.get_tool().")
+        all_tools = self[category]
+        for tool in all_tools:
+            if tool.name == name:
+                return tool
+        raise KeyError(f"Unknown tool '{name}' in category '{category}' "
+                       f"in ToolRepository.")
+
+    def set_default_vendor(self, vendor: str):
```

Comment:
I am happy to have a look/discussion. FWIW, whatr about renaming 'vendor' to 'compiler_suite` ? Then we could have `intel-classic`, and `intel-llvm` (plus `gnu`, `cray`, `nvidia` I'd guess).
Given that otherwise vendor (well. compiler_suite) is a string, any user can add any add whatever they want.

Re other tools - is there a need to have a vendor? If you want (say) specifically a gnu-tar, gnu-make, you can always just create a tool class correspondingly (i.e. either with full path, or with the proper name like `gmake`).